### DISCUSSION TITLE
How to create spacy v3 NER training data in correct format

### DISCUSSION THREAD

**Discussion by tomstelk** [explosion/spaCy#6961] [Help: Coding & Implementations] [answered]
I am trying to train an NER model using spacy v3.
I have pre-labelled training data that might look something like:
[
{
'text': 'ford mondeo 1990',
'entities': [(15, 19, 'year'),  (0, 11, 'car')]
}
]

How can I convert this to the  v3 spacy training binary format? https://spacy.io/api/data-formats#training

It seems like I need to first convert to the json-input format mentioned here: https://spacy.io/api/data-formats#json-input
But this format does not seem to have any where to put entities? 
Is there any way to convert to this json-input format non-manually?

Alternatively can I directly convert text & entities to the binary format without going through this intermediate "json-input" format?

**Comment by ines** [comment, marked as answer]
Hi! I was just going to point you to the data formats docs, but I realised you already linked that page above. As the docs section on the [binary format](https://spacy.io/api/data-formats#training) mentions, it's really just a collection of serialized `Doc` objects under the hood:

> The main data format used in spaCy v3.0 is a binary format created by serializing a `DocBin`, which represents a collection of `Doc` objects. This means that you can train spaCy pipelines using the same format it outputs: annotated `Doc` objects.

The code example on the right of that page shows you how to save out a `.spacy` file. So all you need to do is generate `Doc` objects with your annotations. For example:

```python
doc = nlp.make_doc("ford mondeo 1990")
spans = [doc.char_span(start, end, label=label) for start, end, label in [(12, 16, "year"), (0, 11, "car")]]
doc.ents = [span for spans if span is not None]
```

The check for `None` here is important: in the example you showed above `(15, 19, 'year')` doesn't actually map to a valid character span in the text. So you probably want to catch those cases and review the annotations manually.

> It seems like I need to first convert to the json-input format mentioned here: spacy.io/api/data-formats#json-input

No, that's not true – as the message on that page reads, this is the previous and **deprecated format**. We're only documenting it here for reference in case people still have old files they need to convert.

**Reply by tomstelk** [reply]
Perfect! Thank you very much!

**Reply by austin-e-cox** [reply]
Do either of you know what the procedure would be to do the same for setting doc deps and heads?
They are listed as keyword args to the Doc object, but the approach mentioned in this thread doesn't seem to work for them as they cannot be assigned through doc.deps since there is no doc.deps attribute

Have spent 3+hrs staring at my screen trying different combos out. The only think I think I haven't tried is to explicitly pass them to Doc() instead of trying to generate a doc from nlp and then assign, but then I would need to add in another step where I differentiate the text from the spaces somehow and pass those in as arguments as well since the deps map back to the supplied words argument. 
https://spacy.io/api/doc#init
This is hurting my brain a bit, am I just missing something obvious?

~example incoming data ('This is it', {'heads': [1, 1, 0], 'deps': ['A', 'ROOT', 'B']})

**Reply by austin-e-cox** [reply]
Spent more time on it.
This technically works to get the words and spaces to create a Doc, though I would say it's not pretty since we create a doc via nlp method in order to to determine inputs to subsequently create a doc using Doc.__init__ method.
using nlp = spacy.load("en_core_web_sm")

def get_words_spaces(): ...(below)

words, spaces = get_words_spaces(nlp(text))
--Then you can create a Doc and pass in everything.
doc = Doc(nlp.vocab, words=words, spaces=spaces, deps=deps, heads=heads)

def get_words_spaces(doc):
    words = []
    spaces = []

    # if only 1 token
    if len(doc) == 1:
        words.append(doc[0])
        spaces.append(False)
        return words, spaces

    # loop through all but last token
    for ind, token in enumerate(doc[:-1]):
        words.append(token.text)
        if doc.text[token.idx+len(token):token.idx+len(token)+1] != ' ':
            spaces.append(False)
        else:
            spaces.append(True)

    # add in last token (assumes no space at end of string)
    words.append(str(doc[-1]))
    spaces.append(False)
    return words, spaces

**Reply by adrianeboyd** [reply]
For setting deps by hand, you need to set them on the tokens:

```python
words = ["a", "b", "c"]
deps = ["A", "B", "C"]
heads = [0, 0, 0] # absolute heads
doc = Doc(nlp.vocab, words=words)
for i, (token, dep, head) in enumerate(zip(doc, deps, heads)):
    token.dep_ = dep
    token.head = doc[head] # token.head is a token in the doc, not an integer
```

There's also a built-in helper function `spacy.util.get_words_and_spaces` which converts a list of tokens and a raw text to `words` and `spaces`: https://spacy.io/api/top-level#get_words_and_spaces. In a situation where you have token-based `deps`, you would need to careful about using it, though, because it could insert additional tokens for spaces that throw off the deps offsets. As long as you have at most one space between the tokens in your raw text, it should work, just be aware not to rely on it if there's more whitespace. (Your method above would have similar concerns, so be sure you check what happens for a text like `a b\nc`.)

Internally, we primarily use it for cases where we have `tokens` + `text` + **character-based** entity annotation. The character offsets remain aligned even if there are additional space tokens, so you can use it to create the training `Doc` with gold tokenization AND entity annotation.

**Reply by honnibal** [reply]
> Have spent 3+hrs staring at my screen trying different combos out.

I'm sorry to hear it was frustrating. Out of interest, did you look at the source code? I'm just trying to understand whether you weren't able to find the right place to look, or whether the source didn't really help.

**Reply by austin-e-cox** [reply]
Thanks for the util tip @adrianeboyd.
@honnibal I looked initially on the spacy site at the api documentation, then at the github source to try and sort out my confusion. I believe the confusion arose because with the previous method we had the training data coming in with plain text statements along with the heads/deps marked per token (text, deps=['V','X','ROOT'], heads=[1,2,2]). Since some is coming in plain text and some based on tokens there is some disconnect/mapping assumption. This changes in v3 where everything should come into the trainer token-based (from Examples).

As part of the v3 upgrade it is suggested that the manual training should be abandoned in favor of the cfg file (fair enough).
But it also looks like we are needing Examples instead of the other format, and everything should be wrapped up in docs->docbins to run from the command line. This removes that disconnect of plaintext being mapped to head/dep of tokens during training and pushes it up in the workflow. This means that previous user processes need to be altered/updated as well so everything comes in uniformly as doc data now before any training happens.
My own process was set up to have the same format as spacy 2.x where my statements are plain text but my dep/heads are marked based on assumed tokens (same format as what nlp.update expected) so that needs to change on my end to building docs and mapping the heads/deps in when the docs are built. 

Quick asides:
 - Naming the training object "Example" has made it very difficult to search for it. Any search engine (outside of spacy) just comes up with spacy examples instead of the spacy Example object since they are case insensitive.
- It would be useful to have an example of changing from the previous (text, deps=['V','X','ROOT'], heads=[1,2,2], ner=[]) to using Examples or DocBins

There is another part to why my data was coming in as plain text and that is that I don't otherwise have a great way to view groups of annotation data (data pre-processing). In my file/database structure I can have statements grouped and can quickly view and edit their data. I have not been able to find a good way to visualize groups of statement data with Spacy or Prodigy, only statements individually.
For instance, I may have a selection of 20 short statements that have a similar structure and I want to compare to see if their dep &| head &| ner are all closely matching.
I had purchased prodigy to see if it held any advantages but I cannot seem to find a good way to visualize groups of my data so I honestly haven't been using it and have just been using plain text files/database instead.

Some Background:
When I first started working with spacy my skills/understanding of spacy/nlp were less than stellar. So I had created a number of text files that manually held statements and their dep/head annotations, wrote a parser for the text files, and trained with the output of my parser (the statements and head/dep markings).
I had done it this way so that I could manually adjust any statements quickly and rerun the ~500 test examples to ensure that I was getting results that made at least some sort of sense. These text files gave me a HUD of all my statements deps and heads at once, made it easy to alter any statements, and helped me to quickly run a number of tests to see if the data that I was using would end up with some correlation to what I wanted.
This trial and error was part of how I learned the nlp process.

**Reply by adrianeboyd** [reply]
Thanks for the explanation, it's helpful to see what's confusing. And I realized I misunderstood where you were coming from in your original question.

The closest workflow here, which still isn't quite optimal, is to go through `Example.from_dict`, so you'd have:

```
import spacy
from spacy.training import Example
from spacy.tokens import DocBin

nlp = spacy.blank("en")

example = Example.from_dict(nlp.make_doc(text), {"words": words, "deps": deps, "heads": heads})

db = DocBin()
db.add(example.reference)
db.to_disk("train.spacy")
```

Again, because the spacy tokenization of the text might not align with your reference tokenization, you always want to provide `words` with token-level annotation like `heads`, `deps`, `tags`, etc. For character-based entities or textcat, you can skip `words`, but for the other annotation, you need `words` or the annotation may not get aligned correctly.

**Reply by austin-e-cox** [reply]
Ah that looks far more elegant than what I had written :) I still need to adjust it slightly to get my words to pass into this function, but I think that is just going to be pulling out the make_doc to a variable, getting the tokens from that and passing those in as the words.
The last statement about requiring the ```words``` when words are needed is perfectly understood.

I think the structure of the Example object also was/is confusing me a bit as we start with one data set (text & tags) but are expected to provide two different docs when creating an Example. How are we supposed to have predicted data when we are just initially creating the example? And then from_dict states that it expects predicted doc and gold_standard tags, so the structure from one item but the tags from the other. I just have not been able to get this straight in my head of how I am supposed to have predicted data before I have even started training from my gold standard data. This has led to part of the confusion and time sink. Perhaps I have missed something obvious?

**Reply by adrianeboyd** [reply]
For most common training scenarios, the `predicted` doc stores the predicted tokenization and no other annotation. The predictions are stored in `Example.predicted` during evaluation with `nlp.evaluate`, but internally in a copy so that the evaluation is always run on a clean slate. As a a result, the user won't see those annotations directly.

There are some cases where you might want a layer of gold annotation in the predicted doc that's used when training or evaluating, e.g., to evaluate the entity linker on gold entities, so we want to have the option, but it's not a common use case.

v2 used training tuples with a predicted `Doc` and a reference `GoldParse`. In v3, the annotation from `GoldParse` is moved into a `Doc` so the training annotation and the predicted annotation are all in the exact same format, and the `Example` class stores both along with the token alignment between the two.
